packages/vite/src/node/plugins/resolve.ts

```diff
@@ -659,7 +659,7 @@ export function tryNodeResolve(
   // nearest package.json that may have the `exports` field
   let pkg: PackageData | undefined
 
-  let pkgId = possiblePkgIds.reverse().find((pkgId) => {
```

Comment:
I think that this change is too simple. However, all tests pass - so I've started small.

We probably shouldn't even have a notion of "possible package ids" here - there is only one package id (the root one). Unless a package ID represents an imported source~. Maybe this is relevant for some other things - probably we might still want to look for the nearest package data to determine the scope and thus module format for the given directory. This shouldn't affect `exports` lookup though.

All in all, this probably means that there are not enough tests here - I should add some related to `package.json#type` (for which the "nearest" pkg matters).

Reply:
I don't think we can remove the `reverse()` here. It's used to find the closest `package.json` like you mentioned. And yeah it looks like we lack tests for this.

I thought we had already fixed this issue in https://github.com/vitejs/vite/pull/10371 but looks like this indeed still isn't working. I found that there's a bug in that PR still.

https://github.com/vitejs/vite/blob/2c38bae9458794d42eebd7f7351f5633e2fe8247/packages/vite/src/node/plugins/resolve.ts#L662-L672

Looks like `rootPkgId` should be declared before `pkgId` because `reverse()` mutates `possiblePkgIds`. Making that change fixes it for me too.

Reply:
@bluwy pushed out the proposed fix. Do you have any test cases in mind that would fail with my initial fix? It's probably related to package scopes

Reply:
I also pushed out what I thought would be that test - but it seems that perhaps there are some other bugs in Vite that prevent this from working.

I end up with:
> The requested module '/exports-and-nested-scope/nested-scope/file.js' does not provide an export named 'default'

When I consider the node's resolution semantics though... this fixture package is configured correctly and this should resolve just OK. I've confirmed that with:
```mjs
// playground/resolve/test.mjs

import exportsAndNestedScope from '@vitejs/test-resolve-exports-and-nested-scope/nested'
console.log(exportsAndNestedScope.msg)
```

So I'm not sure how to write this test to verify that the current version of the fix is better than what I had here initially as I'm not sure what exact resolution semantics Vite handles today

Reply:
Thanks for the extra tests. I believe the reason you're getting the error is because the package is linked locally through the `link:` protocol. Linked packages in Vite are treated as source code by default which has to be ESM, so the CJS code in the test package wouldn't work.

I different way to test this could be setting `type: module` in the nested package.json, and change the nested file to ESM instead and so that everything's ESM. I'm not exactly sure if this covers the missing test initially, but I might have a better idea tomorrow after sleep 😅 

For reference, the part was added in https://github.com/vitejs/vite/pull/5665

Reply:
What has been introduced in https://github.com/vitejs/vite/pull/5665 is correct **but** when there are no root exports. `package.json#exports` opt-in the package into the new resolution algorithm and nested `package.json` lookup is not part of that algorithm. After all, with `package.json#exports` you can only import exactly what is being listed in it. The old algorithm supported nested `package.json` lookups though and for packages without `package.json#exports` you have to keep supporting it.

> I different way to test this could be setting type: module in the nested package.json, and change the nested file to ESM instead and so that everything's ESM. I'm not exactly sure if this covers the missing test initially, but I might have a better idea tomorrow after sleep 😅

I don't see how this one would help here. The `package.json#type` is the only thing (that I know of) that can be changed by a nested `package.json` in presence of `exports` in the root's `package.json`. So it's the only observable thing that I can check. 

Maybe I can reverse this situation and put `type: 'module'` in the nested one while keeping `type: 'commonjs'` in the root? Not sure though if I would impact Vite's algorithm enough for this to actually make a difference - I mean, perhaps the nested one would be loaded just OK but not sure if it would fail with the previous version of my fix.

> Thanks for the extra tests. I believe the reason you're getting the error is because the package is linked locally through the link: protocol. Linked packages in Vite are treated as source code by default which has to be ESM, so the CJS code in the test package wouldn't work.

This is a little bit weird. I never really used `link:` myself so maybe I'm missing how this is usually used - but it seems to me that with raise of tools like Turbopack etc it would make sense to allow any package to be treated the same. For perf reasons one could set up transpilation etc for their workspaces that are used by other workspaces through the `link:` and in such a case, they could end up with CJS. 

Either way... take a good sleep! Tomorrow (or later) is as good day as any to fight some resolution issues 😜

Reply:
> What has been introduced in #5665 is correct **but** when there are no root exports. `package.json#exports` opt-in the package into the new resolution algorithm and nested `package.json` lookup is not part of that algorithm. After all, with `package.json#exports` you can only import exactly what is being listed in it. The old algorithm supported nested `package.json` lookups though and for packages without `package.json#exports` you have to keep supporting it.

In retrospect #5665 isn't quite right I think, if you want to get the nearest `package.json` you should resolve the import, then walk up the directory for the closest `package.json`, not testing `foo/bar/package.json` and `foo/package.json`. The first path can be rightly false.

> This is a little bit weird. I never really used `link:` myself so maybe I'm missing how this is usually used - but it seems to me that with raise of tools like Turbopack etc it would make sense to allow any package to be treated the same. For perf reasons one could set up transpilation etc for their workspaces that are used by other workspaces through the `link:` and in such a case, they could end up with CJS.

That's how Vite works currently. Otherwise it's not possible to get HMR when working in a monorepo package.

---

I'd say let's revert the extra tests for now and get the fix in. I think my `type: module` suggestion would make the test pass but wouldn't cover the initial issue too. Thanks for pointing out this section of the code though, the nested `package.json` could use a refactor in a separate PR.

Reply:
> In retrospect https://github.com/vitejs/vite/pull/5665 isn't quite right I think, if you want to get the nearest package.json you should resolve the import, then walk up the directory for the closest package.json, not testing foo/bar/package.json and foo/package.json. The first path can be rightly false.

Yes, you are right - this is true when there are `package.json#exports` in the root `package.json`. The used algorithm is still fine for the "classic" resolution algorithm though, since you can't "remap" paths anyhow.

> I'd say let's revert the extra tests for now and get the fix in. I think my type: module suggestion would make the test pass but wouldn't cover the initial issue too. 

Yeah. I've decided to keep the "flipped" test though - despite the fact that it's passing with either version of the fix etc. It might still be useful to validate your future fixes.

I've pushed out a change, the CI should become green in a moment.